### PR TITLE
fix pod save after s3 multipart upload

### LINKED ISSUE #8072: bug: pod save fails when large lambdas are uploaded. (2.1.1.dev)
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Localstack prints a python stack error when attempting to save a large lambda (>1G) via "pod save".   It references S3, but that is a red-herring.  Once this error is encountered, the pod save exits leaving other services absent from the file.

### Expected Behavior

A services should be saved to the file.

### How are you starting LocalStack?

With the `localstack` script

### Steps To Reproduce

localstack start -d
awslocal lambda create-function  --function-name kinesisToElasticSearch  --zip-file fileb://kinesisToElasticSearch.zip --handler index.handler --runtime nodejs12.x --role arn:aws:iam::000000000000:role/lambda-role --region us-east-1
localstack pod save file://$PWD/localstack.pod

### Environment

```markdown
- OS:MacOS Ventura 13.2.1
- LocalStack: 2.0.1.dev
```

### Anything else?

Note if you delete the lambda, the problem still occurs.   I have not determined the exact limit on the lambda zip file.

Output

2023-04-04T23:51:16.437  WARN --- [-functhread3] hypercorn.error            : ASGI Framework Lifespan error, continuing without Lifespan support
2023-04-04T23:51:16.437  WARN --- [-functhread3] hypercorn.error            : ASGI Framework Lifespan error, continuing without Lifespan support
2023-04-04T23:51:16.438  INFO --- [-functhread3] hypercorn.error            : Running on https://0.0.0.0:4566 (CTRL + C to quit)
2023-04-04T23:51:16.438  INFO --- [-functhread3] hypercorn.error            : Running on https://0.0.0.0:4566 (CTRL + C to quit)
Ready.
2023-04-04T23:52:32.051  INFO --- [   asgi_gw_0] localstack.request.aws     : AWS lambda.CreateFunction => 201
2023-04-04T23:52:32.126  INFO --- [   asgi_gw_2] localstack.request.aws     : AWS sts.AssumeRole => 200
2023-04-04T23:52:43.746  INFO --- [   asgi_gw_2] localstack.request.http    : GET / => 200
2023-04-04T23:52:43.797 ERROR --- [   asgi_gw_0] l.pods.manager             : Error while saving state of service s3 into pod
Traceback (most recent call last):
File "/opt/code/localstack/.venv/lib/python3.10/site-packages/localstack_persistence/pods/manager.py", line 15, in extract_into
try:A.accept_state_visitor(E)
File "/opt/code/localstack/localstack/services/plugins.py", line 145, in accept_state_visitor
ReflectionStateLocator(service=self.name()).accept_state_visitor(visitor)
File "/opt/code/localstack/localstack/state/inspect.py", line 133, in accept_state_visitor
visitor.visit(attribute)
File "/usr/local/lib/python3.10/functools.py", line 926, in _method
return method.__get__(obj, cls)(*args, **kwargs)
File "/opt/code/localstack/.venv/lib/python3.10/site-packages/localstack_persistence/pods/save.py", line 27, in _
G=os.path.join(constants.API_STATES_DIRECTORY,C,B,E,constants.MOTO_BACKEND_STATE_FILE)
File "/usr/local/lib/python3.10/posixpath.py", line 90, in join
genericpath._check_arg_types('join', a, *p)
File "/usr/local/lib/python3.10/genericpath.py", line 152, in _check_arg_types
raise TypeError(f'{funcname}() argument must be str, bytes, or '
TypeError: join() argument must be str, bytes, or os.PathLike object, not 'NoneType'
2023-04-04T23:52:43.805  INFO --- [   asgi_gw_0] localstack.request.http    : GET /_localstack/pods/state => 200
2023-04-04T23:52:44.031  INFO --- [   asgi_gw_2] localstack.request.http    : GET /_localstack/pods/environment => 200

### PR DESCRIPTION
**Pull request by bentsku**
This PR fixes #8072 

Basically, the issue stems from moto accessing thus creating a BackendDict with an `account_id` set to `None`.

Thanks to @taimourfenris, he discovered that it was after a `MultipartUpload`, the issue would arise. It happens also in lambda because lambda makes use of `upload_fileobj`, which will automatically switch to a multipart upload if the file is big.

I tried to debug moto to see when it would access a backend with an null account id.

I've created a small sample to create a multipart:
```python
# https://boto3.amazonaws.com/v1/documentation/api/latest/guide/s3.html#multipart-transfers
chunk_size = 5_242_880 * 10  # around 52MB
with io.BytesIO() as data:
    data.write(os.urandom(chunk_size * 2))
    data.seek(0)
    s3_client.upload_fileobj(data, "test-bucket", "test-key")
```
And I could confirm after running this command that `localstack pod save file://test.zip` would fail. 

Added a print statement to the `__get_item__` of `BackendDict` to see when it would be accessed:
```python
 File "/Users/benjaminsimon/Projects/localstack/localstack/localstack/services/s3/provider.py", line 804, in upload_part
    response: UploadPartOutput = call_moto(context)
  File "/Users/benjaminsimon/Projects/localstack/localstack/localstack/services/moto.py", line 47, in call_moto
    return dispatch_to_backend(context, dispatch_to_moto, include_response_metadata)
  File "/Users/benjaminsimon/Projects/localstack/localstack/localstack/aws/forwarder.py", line 120, in dispatch_to_backend
    http_response = http_request_dispatcher(context)
  File "/Users/benjaminsimon/Projects/localstack/localstack/localstack/services/moto.py", line 115, in dispatch_to_moto
    response = dispatch(request, request.url, request.headers)
  File "/Users/benjaminsimon/Projects/localstack/localstack/localstack/utils/patch.py", line 38, in proxy
    return new(target, *args, **kwargs)
  File "/Users/benjaminsimon/Projects/localstack/localstack/localstack/services/s3/provider.py", line 2043, in _fix_key_response
    status_code, resp_headers, key_value = fn(self, *args, **kwargs)
  File "/Users/benjaminsimon/Projects/localstack/moto/moto/utilities/aws_headers.py", line 74, in _wrapper
    response = f(*args, **kwargs)
  File "/Users/benjaminsimon/Projects/localstack/moto/moto/s3/responses.py", line 1208, in key_response
    response = self._key_response(request, full_url, self.headers)
  File "/Users/benjaminsimon/Projects/localstack/moto/moto/s3/responses.py", line 1320, in _key_response
    return self._key_response_put(request, body, bucket_name, query, key_name)
  File "/Users/benjaminsimon/Projects/localstack/moto/moto/s3/responses.py", line 1516, in _key_response_put
    response_headers.update(key.response_dict)
  File "/Users/benjaminsimon/Projects/localstack/moto/moto/s3/models.py", line 278, in response_dict
    tags = s3_backends[self.account_id]["global"].tagger.get_tag_dict_for_resource( 
```

So the faulty part was there, when returning the part (using a `FakeKey` underneath), here:
https://github.com/getmoto/moto/blob/e0ceec9e48fda482f549ad1f8212731ca82b35a1/moto/s3/responses.py#L1507-L1512

Basically, moto will try to get the `tagger` for the `FakeKey`, but it's just a part instantiated with minimal data, and no account id (because it can't have tag and we don't really care about its account id). 

I've had a look to patch it, but it would pretty tricky as it's contained in `s3/responses.py:S3Response:._key_response_put`, which is around 300LOC. 

The easiest is to directly implement it in our provider and calling directly the moto backend, without then calling `key.response_dict`.

I could validate that it worked with my sample and running `localstack pod save file://test.zip`, it would succeed in creating the pod. I could also load the pod afterwards. I'm not sure how to add a test for this as it's pretty specific.

### CODE REVIEW

File: localstack/services/s3/provider.py [hunk 1]
```diff
@@ -788,14 +788,35 @@ def upload_part(self, context: RequestContext, request: UploadPartRequest) -> Up
...
                 "Part number must be an integer between 1 and 10000, inclusive",
                 ArgumentName="partNumber",
                 ArgumentValue=part_number,
             )
 
-        response: UploadPartOutput = call_moto(context)
+        if body := request.get("Body"):
+            part = body.read()
+        else:
+            part = b""
```
Comment (giograno) [inline localstack/services/s3/provider.py:801]:
```suggestion
        part = body.read() if (body := request.get("Body")) else b""
```

Comment (giograno) [reply localstack/services/s3/provider.py:801]: nit: could probably be shorter

**Review by giograno — APPROVED** [review]
That's just great 💯 thanks for tackling this!

For the persistence part, we could, later on, add a persistence test with a large file to trigger a multipart upload.
